server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -149,6 +168,202 @@ public TransportShardBulkAction(
+        private final PrimaryTermValidationRequest request;
+
+        public PrimaryTermValidationReplicaAction(
+            ActionListener<ReplicaResponse> onCompletionListener,
+            IndexShard replica,
+            ReplicationTask task,
+            PrimaryTermValidationRequest request
+        ) {
+            this.onCompletionListener = onCompletionListener;
+            this.replica = replica;
+            this.task = task;
+            this.request = request;
+        }
+
+        @Override
+        public void onResponse(Releasable releasable) {
+            setPhase(task, "finished");
+            onCompletionListener.onResponse(new ReplicaResponse(SequenceNumbers.NO_OPS_PERFORMED, SequenceNumbers.NO_OPS_PERFORMED));
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            setPhase(task, "failed");
+            onCompletionListener.onFailure(e);
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            setPhase(task, "primary-term-validation");
+            final String actualAllocationId = this.replica.routingEntry().allocationId().getId();
+            if (actualAllocationId.equals(request.getTargetAllocationID()) == false) {
+                throw new ShardNotFoundException(
+                    this.replica.shardId(),
+                    "expected allocation id [{}] but found [{}]",
+                    request.getTargetAllocationID(),
+                    actualAllocationId
+                );
+            }
+            // Check operation primary term against the incoming primary term
+            // If the request primary term is low, then trigger lister failure
+            if (request.getPrimaryTerm() < replica.getOperationPrimaryTerm()) {
+                final String message = String.format(
+                    Locale.ROOT,
+                    "%s operation primary term [%d] is too old (current [%d])",
+                    request.getShardId(),
+                    request.getPrimaryTerm(),
+                    replica.getOperationPrimaryTerm()
+                );
+                onFailure(new IllegalStateException(message));
+            }
+            onResponse(null);
+        }
+    }
+
+    /**
+     * Primary term validation request sent to a specific allocation id
+     *
+     * @opensearch.internal
+     */
+    protected static final class PrimaryTermValidationRequest extends TransportRequest {
+
+        /**
+         * {@link AllocationId#getId()} of the shard this request is sent to
+         **/
+        private final String targetAllocationID;
+        private final long primaryTerm;
+        private final ShardId shardId;
+
+        public PrimaryTermValidationRequest(String targetAllocationID, long primaryTerm, ShardId shardId) {
+            this.targetAllocationID = Objects.requireNonNull(targetAllocationID);
+            this.primaryTerm = primaryTerm;
+            this.shardId = Objects.requireNonNull(shardId);
+        }
```

Comment:
Can we have both source primary shard routing and replica shard routing details, I guess we only have target shard details

Reply:
While this can be added, is it really required? In the status quo currently, across all replication requests, we do not send source allocation id. Do you see any usecases where the source allocation id would be required? Right now, I have added variables which are absolutely required.